### ISSUE TITLE
TensorRT .engine model shows minimal practical performance gain: higher GPU memory usage, subtle latency impact despite faster inference

### ISSUE THREAD

**Issue by myungki-park** [ultralytics/ultralytics#21152]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

## ❓ Question / Feedback

**I tested YOLOv11 `.pt` vs `.engine` (TensorRT) models in a real-world multi-stage inference pipeline (3 stages, 2000+ images)** and I want to share practical findings and limitations.

---

## 🧠 Context

- Model: YOLOv11 (segmentation and classification tasks)
- Exported using `model.export(format='engine', half=False, dynamic=False, imgsz=640, batch=1) for 1st and 2nd segmentation model / model.export(format='engine', half=False, dynamic=True, imgsz=224, batch=64) for 3rd Classification Model
- For Inference Phase, Load Model with designated task as model(engine_path, task='segment')
- Inference run: ~2000+ image test set
- System: NVIDIA A-100 GPU, CUDA12.2, Python 3.10, Ultralytics latest release

- With the Half=True option when export 'engine', belows are the equivalnece check results, and the equivalnce between .pt and .engine is acceptable. 

![Image](https://github.com/user-attachments/assets/367bfccd-dc35-43b4-98a1-88094e33267f)

---

## 🔁 My Pipeline (3-stage)

1. **Stage 1**: Segmentation objcect on image
2. **Stage 2**: Segmentation sub-object on object
3. **Stage 3**: Do Classification

All models were tested in both `.pt` and `.engine` form.

---

## 📊 What I expected from TensorRT:
- Lower GPU memory usage
- Lower end-to-end latency
- Lower CUDA kernel count

---

## 😓 What I actually observed:

| Metric               | Observation |
|----------------------|-------------|
| Inference latency    | ✅ *Significantly lower per model* |
| End-to-end latency   | ❌ *Improvement is minimal* (inference is a small % of pipeline time) |
| GPU memory usage     | ❌ *.engine uses more GPU memory* (workspace allocation?) |
| CUDA kernel count    | ❌ *Only slightly lower* |
| Final gain           | ❌ *Negligible impact on actual pipeline throughput* |

*Belows are the codes for each metric

**<CUDA KERNAL>**
from pynvml import (
    nvmlInit, nvmlShutdown, nvmlDeviceGetHandleByIndex, nvmlDeviceGetUtilizationRates,
    NVMLError
)

try:
    nvmlInit()
    _NVML_AVAILABLE = True
except NVMLError:
    _NVML_AVAILABLE = False
    logging.warning("NVML initialization failed. GPU utilization will not be logged.")

gpu_util = 0
        if torch.cuda.is_available() and _NVML_AVAILABLE:
            try:
                handle = nvmlDeviceGetHandleByIndex(0)  # GPU 0
                util = nvmlDeviceGetUtilizationRates(handle)
                gpu_util = util.gpu  # %
            except NVMLError as nv_err:
                logging.warning(f"Failed to fetch GPU utilization: {nv_err}")

**<GPU MEM>**
torch.cuda.reset_peak_memory_stats()
torch.cuda.synchronize()
mem_allocated = torch.cuda.max_memory_allocated() / (1024 ** 2)  # MB

![Image](https://github.com/user-attachments/assets/5a217594-439e-4357-beeb-6379a7ba7cb5)

![Image](https://github.com/user-attachments/assets/67994aa1-082d-43c9-834a-82cf6fba2be7)

![Image](https://github.com/user-attachments/assets/949e2e76-2b82-44c9-b32f-7b184a2d4e8a)

---

## 🔍 Notes:

- All 3 stages use `.engine` models
- Inference acceleration is clear in isolation, but **pre/post-processing dominates total runtime**
- **TensorRT workspace preallocation may be causing higher memory usage**
- I suspect **low batch size** in Stage 1/2 and **CPU-heavy crop/postprocess** are bottlenecks

---

## 🙋 My Question / Suggestion:

- Is this expected behavior?
- Could Ultralytics improve:
  - memory usage when exporting `.engine`?
  - integration of full-pipeline batching?
  - documentation on when TensorRT is actually beneficial?

This kind of practical profiling info might help users avoid overoptimizing the wrong parts of their pipeline.

Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @myungki-park, thank you for sharing your detailed findings and feedback with Ultralytics 🚀! Your profiling insights and suggestions are extremely valuable to the ongoing development and documentation of YOLO and the export process.

This is an automated response to help streamline support, but an Ultralytics engineer will also review your report and assist you further soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further, especially if you suspect unexpected memory usage or performance in a specific workflow.

If you are sharing a custom training or deployment ❓ Question, please include as much detail as possible—such as logs, code snippets, or resource utilization traces—to help us understand and address your observations. For optimizing pipeline performance, please ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) and review the latest [export documentation](https://docs.ultralytics.com/modes/export/) for any recent updates.

Join the Ultralytics community for real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧, in-depth discussions on [Discourse](https://community.ultralytics.com/), or knowledge sharing on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Please ensure you are using the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your contribution and for helping to improve YOLO and Ultralytics HUB! 🧡

**Comment by Y-T-G** [comment]
What's your inference code?

**Comment by myungki-park** [comment]
@Y-T-G 
for the inference, this is my model class 

![Image](https://github.com/user-attachments/assets/44b1f455-0c47-4142-b44a-257f8bd76e2e)

for each inference stage, inference code is 

![Image](https://github.com/user-attachments/assets/19e1565b-81b1-4afa-9399-b263afeb4be9)

**Comment by Y-T-G** [comment]
You should export to engine with `half=True` and `nms=True`

**Comment by myungki-park** [comment]
@Y-T-G I'm sorry but with the nms=True option i got an error 

[TRT] [E] 1: [jitConvRunner.cpp::nvinfer1::rt::task::CaskJitConvRunner::execute::378] Error Code 1: Cask (Failed to run the convolution kernel!)

**Comment by Y-T-G** [comment]
I guess just use `half=True` then

**Comment by myungki-park** [comment]
@Y-T-G hah... i alreay export .engine with the option half=True....
